### ISSUE TITLE
Gaussian Kernel python interface

### ISSUE THREAD

**Issue by jomsdev** [xdata-skylark/libskylark#20]
I am working on the python bindings in the capi branch. I wrote a little script which read a data file and applies the Gaussian kernel to the data. The schema would be something like that

```python
import skylark.io as sl_io
import skylark.ml.kernels as sl_kernels
import El

# Read the file
X = El.DistMatrix()
L = El.DistMatrix()
sl_io.readlibsvm('data/usps.t', X, L, 0)

# Apply the kernel
K = El.DistMatrix()
Gaussian_kernel = sl_kernels.Gaussian(X.Height(), 10.0)
Gaussian_kernel.gram(X, K)
```

The C++ code throws an exception with the message  **This combination has not yet been implemented for Gram**. I tried it reading the data as "rows" and as "columns" but, in both cases, the result has been the same.

Following the execution, the first C++ code to execute the gram functions is in the [ckernel.cpp](https://github.com/xdata-skylark/libskylark/blob/capi/capi/ckernel.cpp#L84) and it calls to the Gram functions, coded in [kernels.hpp](https://github.com/xdata-skylark/libskylark/blob/capi/ml/kernels.hpp#L78). I finally understood how is the flow of that function (Should we make it easier?? With a macro, the returns that end the function inside the macro, etc..).  

The problem could be:

*  SKYLARK_NO_ANY is defined and then C++ directly throw the exception.
*  SKYLARK_NO_ANY is NOT DEFINED but  **[SKYLARK_GRAM_ANY_APPLY_DISPATCH ](https://github.com/xdata-skylark/libskylark/blob/capi/ml/kernels.hpp#L82)** macro is not matching any of the if/else cases.

I am not being able to reproduce the error with the bad results (NaN, etc...) in the python matrix like @haimav . Are they happening in the CAPI branch?

**Comment by jomsdev** [comment]
I have been debugging the C++ code to figure out what is happening. When I call the python function passing DistMatrix() objects as parameters the [SKYLARK_GRAM_ANY_APPLY_DISPATCH](https://github.com/xdata-skylark/libskylark/blob/capi/ml/kernels.hpp#L82) macro is not matching any value. I used that snippet to check te types of our variables:

```cpp
std::string msg = "This combination has not yet been implemented for Gram\n";

    msg = msg + " X: " + X.type().name() + '\n';
    msg = msg + " mdtypes::matrix_t: " + typeid(mdtypes::matrix_t*).name() + '\n';
    msg = msg + " mdtypes::el_matrix_t: " + typeid(mdtypes::el_matrix_t*).name() + '\n';
    msg = msg + " mdtypes::mftypes::matrix_t: " + typeid(mftypes::matrix_t*).name() + '\n';
    msg = msg + " mdtypes::mftypes::el_matrix_t: " + typeid(mftypes::el_matrix_t*).name() + '\n';
```

I pass the msg variable as the default msg for the SKYLARK_THROW_EXCEPTION.
The output obtined is:

```cpp
X: PN2El10DistMatrixIdLNS_6DistNS4DistE0ELS2_2ELNS_10DistWrapNS8DistWrapE0EEE
mdtypes::matrix_t: PN2El6MatrixIdEE
mdtypes::el_matrix_t: PN2El15ElementalMatrixIdEE
mdtypes::mftypes::matrix_t: PN2El6MatrixIfEE
mdtypes::mftypes::el_matrix_t: PN2El15ElementalMatrixIfEE
```

Summarizing, X.type() is different to all the types that we are comparing with it.

Am I doing something wrong or misunderstanding something? Which would be the better approach to solve this?

**Comment by jomsdev** [comment]
Problem solved in the  #21 commit.

The Kernel is working now when is called from the python interface.
